sympy/core/basic.py

```diff
@@ -1624,31 +1624,75 @@ def count_ops(self, visual=None):
         return count_ops(self, visual)
 
     def doit(self, **hints):
-        """Evaluate objects that are not evaluated by default like limits,
-        integrals, sums and products. All objects of this kind will be
-        evaluated recursively, unless some species were excluded via 'hints'
-        or unless the 'deep' hint was set to 'False'.
+        """Evaluate objects that are not evaluated like operations, limits,
+        integrals, sums and products.
 
-        >>> from sympy import Integral
+        Explanation
+        ===========
+
+        Arguments will be evaluated recursively, unless some species were
+        excluded via 'properties' or unless the 'deep' hint was set to 'False'.
+        Evaluatable classes should define their own `_eval_derivative` method,
+        which returns evaluated result. In this case, non-recursive evaluation
+        should also be done when `deep=False` is passed. It is also encouraged
+        to allow `evaluate` option to the constructor.
+
+        Examples
+        ========
+
+        >>> from sympy import Add, Integral, Derivative
         >>> from sympy.abc import x
 
-        >>> 2*Integral(x, x)
-        2*Integral(x, x)
+        >>> expr = Add(Integral(x, x), Integral(x, x), Derivative(x, x), evaluate=False)
+        >>> expr
+        Derivative(x, x) + Integral(x, x) + Integral(x, x)
 
-        >>> (2*Integral(x, x)).doit()
-        x**2
+        >>> expr.doit()
+        x**2 + 1
 
-        >>> (2*Integral(x, x)).doit(deep=False)
-        2*Integral(x, x)
+        >>> expr.doit(deep=False)
+        Derivative(x, x) + 2*Integral(x, x)
+
+        See Also
+        ========
 
+        core.operations.AssocOp
         """
-        if hints.get('deep', True):
-            terms = [term.doit(**hints) if isinstance(term, Basic) else term
-                                         for term in self.args]
+        deep = hints.get('deep', True)
+        properties = hints.get('properties', [])
+
+        eval_doit = self._eval_doit(**hints)
+        if eval_doit is not None:
+            return eval_doit
+        if deep:
+            if any(not f(self) for f in properties):
+                return self
+            terms = [term.doit(**hints) if isinstance(term, Basic)
+                        else term for term in self.args]
             return self.func(*terms)
         else:
             return self
 
+    def _eval_doit(self, **hints):
+        # May be overridden in subclasses
+        return None
+
+    def _eval_doit_evaluatable(self, **hints):
+        """Pre-generated method that can be used as `_eval_doit`
+        for evaluatable classes, i.e. Add, Mul, etc.
+        """
+        deep = hints.get('deep', True)
+        properties = hints.get('properties', [])
```

Comment:
Why is this logic duplicated in `_eval_doit_evaluatable` and `doit`. Surely it only needs to be in the `doit` method.

Reply:
Oops. Thanks.